### PR TITLE
968 fix unprotected map access in monitoring

### LINKED ISSUE #968: [apiusagemonitoring] Unprotected access to map causes panic
Here: https://github.com/zalando/skipper/pull/950/files#diff-a3889b1942785e41a55bf77d4653e68cR103

### PR DESCRIPTION
**Pull request by maxim-tschumak**
Fixes #968

### CODE REVIEW

File: filters/apiusagemonitoring/pathinfo.go [hunk 1]
```diff
@@ -16,8 +17,9 @@ type pathInfo struct {
 	CommonPrefix   string
 	ClientPrefix   string
 
-	metricPrefixesPerMethod [methodIndexLength]*endpointMetricNames
-	metricPrefixedPerClient map[string]*clientMetricNames
+	metricPrefixesPerMethod [methodIndexLength]*endpointMetricNames // endpoint metric names cache
+	metricPrefixedPerClient map[string]*clientMetricNames           // client metric names cache
```
Comment (tkrop) [inline filters/apiusagemonitoring/pathinfo.go]: Let us use `sync.Map` as replacement here?

The code to maintain will be shorter and easier to understand using a standard pattern, and there are smart guys that have thought about the performance implications of locks and formed it into this code.

Comment (aryszka) [reply filters/apiusagemonitoring/pathinfo.go]: I would rather try to put the endpointMetricNames and the clientMetricNames into a single combined structure and protect it with a single RW lock scope, if the logic allows. As an alternative approach, I would even consider putting the scope of the Response() method of the filter into a single shared RW lock scope. The reason why the latter may be better, is that the endpointMetricNames is always accessed, and this way the code can be safer for future changes. For performance optimizations, we should create benchmarks first.

File: filters/apiusagemonitoring/filter_clientmetrics_test.go [hunk 2]
```diff
@@ -215,7 +218,7 @@ func Test_Filter_ClientMetrics_NoAuthHeader(t *testing.T) {
 		realmKeyName:          "realm",
 		clientKeyName:         "client",
 		clientTrackingPattern: clientTrackingPatternJustSomeUsers,
-		header:                http.Header{
+		header: http.Header{
```
Comment (szuecs) [inline filters/apiusagemonitoring/filter_clientmetrics_test.go]: please run `make fmt`

File: filters/apiusagemonitoring/pathinfo.go (hunk 1 above)
Comment (szuecs) [reply filters/apiusagemonitoring/pathinfo.go]: sync.Map is not simpler to maintain. 
+1 for @aryszka suggestion

**Comment by maxim-tschumak** [comment]
replaced by #973

**Pull request closed without merging**